src/mgmt/srv_system.c

```diff
@@ -0,0 +1,848 @@
+		if (rc == 0) {
+			D_ERROR("rank %u requested by server "DF_UUID
+				" already taken\n", in->ji_rank,
+				DP_UUID(in->ji_server.sr_uuid));
+			rc = -DER_EXIST;
+			goto out_lock;
+		} else if (rc != -DER_NONEXIST) {
+			D_ERROR("failed to verify rank for server "DF_UUID
+				": %d\n", DP_UUID(in->ji_server.sr_uuid), rc);
+			goto out_lock;
+		}
+		rank = in->ji_rank;
+	}
+
+	rc = add_server(&tx, svc, rank, &in->ji_server);
+	if (rc != 0) {
+		D_ERROR("failed to add server "DF_UUID" as rank %u: %d\n",
+			DP_UUID(in->ji_server.sr_uuid), rank, rc);
+		goto out_lock;
+	}
+
+	map_version = svc->ms_map_version + 1;
+	daos_iov_set(&value, &map_version, sizeof(map_version));
+	rc = rdb_tx_update(&tx, &svc->ms_root, &ds_mgmt_prop_map_version,
+			   &value);
+	if (rc != 0) {
+		D_ERROR("failed to increment map version to %u: %d\n",
+			map_version, rc);
+		goto out_lock;
+	}
+
+	rc = rdb_tx_commit(&tx);
+	if (rc != 0) {
+		D_ERROR("failed to commit map version %u: %d\n", map_version,
+			rc);
+		goto out_lock;
+	}
+
+	D_DEBUG(DB_TRACE, "rank %u joined in map version %u\n", rank,
+		map_version);
+	svc->ms_map_version = map_version;
+	if (in->ji_rank != -1)
+		svc->ms_rank_next = rank_next;
+	notify_map_distributord(svc);
+	out->jo_rank = rank;
+	out->jo_flags = SERVER_IN;
+
+out_lock:
+	ABT_rwlock_unlock(svc->ms_lock);
+	rdb_tx_end(&tx);
+out_svc:
+	mgmt_svc_put_leader(svc);
+out:
+	return rc;
+}
+
+static int
+map_update_bcast(crt_context_t ctx, struct mgmt_svc *svc, uint32_t map_version,
+		 int nservers, struct server_entry servers[])
+{
+	struct mgmt_tgt_map_update_in  *in;
+	struct mgmt_tgt_map_update_out *out;
+	crt_opcode_t			opc;
+	crt_rpc_t		       *rpc;
+	int				rc;
+
+	D_DEBUG(DB_MGMT, "enter: version=%u nservers=%d\n", map_version,
+		nservers);
+
+	opc = DAOS_RPC_OPCODE(MGMT_TGT_MAP_UPDATE, DAOS_MGMT_MODULE, 1);
+	rc = crt_corpc_req_create(ctx, NULL /* grp */,
+				  NULL /* excluded_ranks */, opc,
+				  NULL /* co_bulk_hdl */, NULL /* priv */,
+				  0 /* flags */,
+				  crt_tree_topo(CRT_TREE_KNOMIAL, 32), &rpc);
+	if (rc != 0) {
+		D_ERROR("failed to create system map update RPC: %d\n", rc);
+		goto out;
+	}
+	in = crt_req_get(rpc);
+	in->tm_servers.ca_count = nservers;
+	in->tm_servers.ca_arrays = servers;
+	in->tm_map_version = map_version;
+
+	rc = dss_rpc_send(rpc);
+	if (rc != 0)
+		goto out_rpc;
+
+	out = crt_reply_get(rpc);
+	if (out->tm_rc != 0)
+		rc = -DER_IO;
+
+out_rpc:
+	crt_req_decref(rpc);
+out:
+	D_DEBUG(DB_MGMT, "leave: version=%u nservers=%d: %d\n", map_version,
+		nservers, rc);
+	return rc;
+}
+
+static int
+distribute_map(crt_context_t ctx, struct mgmt_svc *svc)
+{
+	struct rdb_tx		tx;
+	struct enum_server_arg	arg = {};
+	int			rc;
+
+	rc = rdb_tx_begin(svc->ms_rsvc.s_db, svc->ms_rsvc.s_term, &tx);
+	if (rc != 0)
+		goto out;
+
+	ABT_rwlock_rdlock(svc->ms_lock);
+
+	rc = rdb_tx_iterate(&tx, &svc->ms_servers, false /* !backward */,
+			    enum_server_cb, &arg);
+	if (rc != 0)
+		goto out_lock;
```

Comment:
Just curious, how many servers here? it has to send by one RPC?

Reply:
A system can have a couple of thousand servers. I think you have a valid concern; it just would make the implement way more complex if we were to support sending chunk by chunk.